File: measure.py
#!/usr/bin/env python3
import smbus2
import time
import csv
import sys
from datetime import datetime
from adxl345 import ADXL345, OutputDataRate

def init_adxl(adxl):
    """Initialize ADXL345 for continuous measurement with FIFO."""
    print("\n=== ADXL345 INITIALIZATION ===")

    # Set FIFO to STREAM mode
    adxl.fifo_ctl.write("MODE", 0b10)  # STREAM mode

    # Enable measurement mode
    print("Enabling MEASUREMENT mode...")
    adxl.power_control.write("MEASURE", 1)

    # Verify settings
    print("\n*** REGISTER VERIFICATION ***")
    print(f"POWER_CTL.MEASURE:    {adxl.power_control.read('MEASURE')}")
    print(f"BW_RATE.RATE:         0x{adxl.bandwidth_rate.read('RATE'):02X}")
    print(f"FIFO_CTL.MODE:        0b{adxl.fifo_ctl.read('MODE'):02b}")
    print(f"FIFO_CTL.SAMPLES:     {adxl.fifo_ctl.read('SAMPLES')}")
    print(f"FIFO_STATUS.ENTRIES:  {adxl.fifo_status.read('ENTRIES')}")

    print("\nInitialization complete.\n")


def draw_fifo_bar(num_entries, watermark=28, max_size=32, bar_width=40):
    """Draw a visual representation of FIFO fill level.
    This is SLOW. Was using it for fun but you can't achieve very fast
    ODRs without overflow here.

    Args:
        num_entries: Current number of samples in FIFO
        watermark: Watermark level
        max_size: Maximum FIFO size
        bar_width: Width of the bar in characters

    Returns:
        str: Visual bar representation
    """
    # Calculate how many characters to fill
    fill_chars = int((num_entries / max_size) * bar_width)
    watermark_pos = int((watermark / max_size) * bar_width)

    # Build the bar
    bar = ""
    for i in range(bar_width):
        if i < fill_chars:
            bar += "█"
        elif i == watermark_pos:
            bar += "|"  # Show watermark position
        else:
            bar += "░"

    return f"[{bar}] {num_entries:2d}/32"


def read_continuous(adxl, duration_seconds=10):
    """Continuously read FIFO with watermark monitoring to avoid data loss.

    Args:
        adxl: ADXL instance
        duration_seconds: How long to acquire data
        sample_rate: Expected sample rate in Hz

    Returns:
        list: List of (timestamp, x_g, y_g, z_g) tuples
    """
    sample_rate = adxl.odr.hz
    sample_period = 1.0 / sample_rate
    watermark = adxl.watermark

    print(f"=== STARTING CONTINUOUS ACQUISITION ===")
    print(f"Duration: {duration_seconds}s")
    print(f"Sample rate: {sample_rate} Hz")
    print(f"Expected samples: ~{duration_seconds * sample_rate}")
    print(f"Watermark set at {adxl.watermark} samples")
    print(f"Reading when samples available...\n")

    overflow_count = 0
    read_count = 0
    samples = []

    # Read INT_SOURCE register to clear overrun and watermark bits
    watermark_flag = adxl.interrupt_source.read("WATERMARK")
    time.sleep(0.01)

    start_time = time.time()

    while time.time() - start_time < duration_seconds:
        watermark_reached = adxl.interrupt_source.read("WATERMARK")
        overflow_occured = adxl.interrupt_source.read("OVERRUN")

        if overflow_occured:
            overflow_count += 1

        # Read waterflow amount of samples
        if watermark_reached:
            for _ in range(watermark):
                samples.append(adxl.get_accel())
            read_count += watermark

    # Create timestamped samples
    timestamped_samples = []
    for i, (x, y, z) in enumerate(samples):
        timestamp = i * sample_period
        timestamped_samples.append((timestamp, x, y, z))

    print(f"=== ACQUISITION COMPLETE ===")
    print(f"Collected {len(timestamped_samples)} samples in {duration_seconds}s")
    print(f"Expected: ~{duration_seconds * sample_rate}")
    print(f"Overflow events: {overflow_count}")
    print(f"Read operations: {read_count}\n")

    return timestamped_samples


def write_to_csv(samples, settings, filename=None):
    """Write samples with timestamps and metadata to CSV file.

    Args:
        samples: List of (timestamp, x_g, y_g, z_g) tuples
        filename: Output filename (auto-generated if None)

    Returns:
        str: The filename written to
    """
    if filename is None:
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        filename = fr"data/accelerometer_data_{timestamp}.csv"

    print(f"Writing {len(samples)} samples to {filename}...")

    with open(filename, 'w', newline='') as csvfile:
        writer = csv.writer(csvfile)

        # Write metadata (settings)
        writer.writerow(settings)
        # Write header
        writer.writerow(['time_s', 'x_g', 'y_g', 'z_g'])

        # Write data
        for timestamp, x_g, y_g, z_g in samples:
            writer.writerow([
                f"{timestamp:.6f}",
                f"{x_g:.6f}",
                f"{y_g:.6f}",
                f"{z_g:.6f}"
            ])

    return filename


def print_sample_preview(samples, num_preview=10):
    """Print first and last few samples as preview."""
    if not samples:
        print("No samples to preview.")
        return

    print("=== SAMPLE PREVIEW ===")
    print(f"{'Time(s)':<12} {'X(g)':<10} {'Y(g)':<10} {'Z(g)':<10}")
    print("-" * 42)

    # First samples
    for timestamp, x, y, z in samples[:num_preview]:
        print(f"{timestamp:<12.6f} {x:<10.4f} {y:<10.4f} {z:<10.4f}")

    if len(samples) > num_preview * 2:
        print("...")

        # Last samples
        for timestamp, x, y, z in samples[-num_preview:]:
            print(f"{timestamp:<12.6f} {x:<10.4f} {y:<10.4f} {z:<10.4f}")

    print()

def flush(adxl):
    adxl.fifo_ctl.write("MODE", 0b00)  # BYPASS mode
    num_entries = adxl.fifo_status.read("ENTRIES")
    while num_entries != 0:
        adxl.get_accel()

def terminate(adxl):
    # Sets sensor to BYPASS mode so FIFO does not fill
    adxl.fifo_ctl.write("MODE", 0b00)  # BYPASS mode

def measure(sensor: ADXL345, is_stationary : bool = False):
    """
    Main measurement routine.
        sensor: ADXL345 object allowing register operations.
        is_stationary: Whether this is an active or passive run.
    """
    init_adxl(sensor)

    # Acquisition parameters
    duration = 10 # seconds

    settings = [
        f"Stationary: {is_stationary}",
        f"Output Data Rate:{sensor.odr.hz} Hz ",
        f"Range: {sensor.g_range.g} g"
    ]

    # Perform continuous acquisition
    samples = read_continuous(sensor, duration_seconds=duration)

    # Write to CSV
    filename = write_to_csv(samples, settings)

    # Print preview
    print_sample_preview(samples, num_preview=10)

    print("=" * 57)
    print(f"Measurement complete!")
    print(f"Data saved to: {filename}")
    print("=" * 57)
    print()

    terminate(sensor)


if __name__ == "__main__":
    try:
        bus = smbus2.SMBus(1)
        sensor = ADXL345(
            0x53,
            bus,
            watermark=28,
            odr=OutputDataRate.ODR_100
        )

        measure(sensor)
    except KeyboardInterrupt:
        print("\n\nMeasurement interrupted by user.")
    except Exception as e:
        print(f"\n\nError: {e}")
        import traceback
        traceback.print_exc()

File: measure.cpp
#include "measure.hpp"
#include "adxl345.hpp"

#include <iostream>
#include <fstream>
#include <iomanip>
#include <sstream>
#include <chrono>
#include <ctime>
#include <stdexcept>
#include <algorithm>

// ---------------------------------------------------------------------------
// init_adxl
// ---------------------------------------------------------------------------
void init_adxl(ADXL345& adxl) {
    std::cout << "\n=== ADXL345 INITIALIZATION ===\n";

    // Set FIFO to Stream mode (0b10)
    adxl.fifo_ctl.write("MODE", 0b10);

    // Enable measurement mode
    std::cout << "Enabling MEASUREMENT mode...\n";
    adxl.power_control.write("MEASURE", 1);

    // Verify settings
    std::cout << "\n*** REGISTER VERIFICATION ***\n";
    std::cout << "POWER_CTL.MEASURE:    "
              << static_cast<int>(adxl.power_control.read("MEASURE")) << "\n";
    std::cout << "BW_RATE.RATE:         0x"
              << std::hex << std::setw(2) << std::setfill('0')
              << static_cast<int>(adxl.bandwidth_rate.read("RATE"))
              << std::dec << "\n";
    std::cout << "FIFO_CTL.MODE:        0b"
              << static_cast<int>(adxl.fifo_ctl.read("MODE")) << "\n";
    std::cout << "FIFO_CTL.SAMPLES:     "
              << static_cast<int>(adxl.fifo_ctl.read("SAMPLES")) << "\n";
    std::cout << "FIFO_STATUS.ENTRIES:  "
              << static_cast<int>(adxl.fifo_status.read("ENTRIES")) << "\n";

    std::cout << "\nInitialization complete.\n\n";
}

// ---------------------------------------------------------------------------
// read_continuous
// ---------------------------------------------------------------------------
std::vector<Sample> read_continuous(ADXL345& adxl, double duration_seconds) {
    double sample_rate  = odr_to_hz(adxl.odr);
    double sample_period = 1.0 / sample_rate;
    int    watermark    = adxl.watermark;

    std::cout << "=== STARTING CONTINUOUS ACQUISITION ===\n";
    std::cout << "Duration:         " << duration_seconds << " s\n";
    std::cout << "Sample rate:      " << sample_rate << " Hz\n";
    std::cout << "Expected samples: ~"
              << static_cast<long>(duration_seconds * sample_rate) << "\n";
    std::cout << "Watermark set at: " << watermark << " samples\n";
    std::cout << "Reading when samples available...\n\n";

    int overflow_count = 0;
    int read_count     = 0;

    std::vector<Sample> raw_samples;
    // Reserve 10 % extra capacity to absorb timing jitter without reallocation.
    constexpr double RESERVE_FACTOR = 1.1;
    raw_samples.reserve(static_cast<size_t>(duration_seconds * sample_rate * RESERVE_FACTOR));

    // Clear interrupt flags before starting
    adxl.interrupt_source.read("WATERMARK");

    using Clock = std::chrono::steady_clock;
    auto start  = Clock::now();
    auto end    = start + std::chrono::duration<double>(duration_seconds);

    while (Clock::now() < end) {
        bool watermark_reached = adxl.interrupt_source.read("WATERMARK") != 0;
        bool overflow_occurred = adxl.interrupt_source.read("OVERRUN")   != 0;

        if (overflow_occurred) {
            ++overflow_count;
        }

        if (watermark_reached) {
            for (int i = 0; i < watermark; ++i) {
                auto [x, y, z] = adxl.get_accel();
                raw_samples.push_back({0.0, x, y, z});
            }
            read_count += watermark;
        }
    }

    // Assign evenly-spaced timestamps based on the configured ODR
    std::vector<Sample> timestamped;
    timestamped.reserve(raw_samples.size());
    for (size_t i = 0; i < raw_samples.size(); ++i) {
        timestamped.push_back({
            static_cast<double>(i) * sample_period,
            raw_samples[i].x,
            raw_samples[i].y,
            raw_samples[i].z
        });
    }

    std::cout << "=== ACQUISITION COMPLETE ===\n";
    std::cout << "Collected " << timestamped.size()
              << " samples in " << duration_seconds << " s\n";
    std::cout << "Expected: ~"
              << static_cast<long>(duration_seconds * sample_rate) << "\n";
    std::cout << "Overflow events:  " << overflow_count << "\n";
    std::cout << "Read operations:  " << read_count << "\n\n";

    return timestamped;
}

// ---------------------------------------------------------------------------
// write_to_csv
// ---------------------------------------------------------------------------
std::string write_to_csv(const std::vector<Sample>& samples,
                         const std::vector<std::string>& settings,
                         const std::string& filename) {
    std::string out_file = filename;
    if (out_file.empty()) {
        // Generate timestamped filename
        auto now      = std::chrono::system_clock::now();
        std::time_t t = std::chrono::system_clock::to_time_t(now);
        std::tm     tm_info;
        localtime_r(&t, &tm_info);
        char buf[32];
        std::strftime(buf, sizeof(buf), "%Y%m%d_%H%M%S", &tm_info);
        out_file = std::string("data/accelerometer_data_") + buf + ".csv";
    }

    std::cout << "Writing " << samples.size()
              << " samples to " << out_file << "...\n";

    std::ofstream csv(out_file);
    if (!csv.is_open()) {
        throw std::runtime_error("Failed to open output file: " + out_file);
    }

    // Metadata row
    for (size_t i = 0; i < settings.size(); ++i) {
        if (i > 0) csv << ',';
        csv << settings[i];
    }
    csv << '\n';

    // Header row
    csv << "time_s,x_g,y_g,z_g\n";

    // Data rows
    csv << std::fixed << std::setprecision(6);
    for (const auto& s : samples) {
        csv << s.timestamp << ','
            << s.x        << ','
            << s.y        << ','
            << s.z        << '\n';
    }

    return out_file;
}

// ---------------------------------------------------------------------------
// print_sample_preview
// ---------------------------------------------------------------------------
void print_sample_preview(const std::vector<Sample>& samples, int num_preview) {
    if (samples.empty()) {
        std::cout << "No samples to preview.\n";
        return;
    }

    std::cout << "=== SAMPLE PREVIEW ===\n";
    std::cout << std::left
              << std::setw(12) << "Time(s)"
              << std::setw(10) << "X(g)"
              << std::setw(10) << "Y(g)"
              << std::setw(10) << "Z(g)" << '\n';
    std::cout << std::string(42, '-') << '\n';

    auto print_row = [](const Sample& s) {
        std::cout << std::fixed << std::setprecision(6)
                  << std::left << std::setw(12) << s.timestamp
                  << std::setprecision(4)
                  << std::setw(10) << s.x
                  << std::setw(10) << s.y
                  << std::setw(10) << s.z << '\n';
    };

    int n = static_cast<int>(samples.size());
    int show_head = std::min(num_preview, n);

    for (int i = 0; i < show_head; ++i) {
        print_row(samples[static_cast<size_t>(i)]);
    }

    if (n > num_preview * 2) {
        std::cout << "...\n";
        for (int i = n - num_preview; i < n; ++i) {
            print_row(samples[static_cast<size_t>(i)]);
        }
    }

    std::cout << '\n';
}

// ---------------------------------------------------------------------------
// flush
// ---------------------------------------------------------------------------
void flush(ADXL345& adxl) {
    adxl.fifo_ctl.write("MODE", 0b00);  // Bypass — stops FIFO
    uint8_t entries = adxl.fifo_status.read("ENTRIES");
    while (entries != 0) {
        adxl.get_accel();
        entries = adxl.fifo_status.read("ENTRIES");
    }
}

// ---------------------------------------------------------------------------
// terminate
// ---------------------------------------------------------------------------
void terminate(ADXL345& adxl) {
    adxl.fifo_ctl.write("MODE", 0b00);  // Bypass — FIFO stops collecting
}

// ---------------------------------------------------------------------------
// measure
// ---------------------------------------------------------------------------
void measure(ADXL345& sensor, bool is_stationary) {
    init_adxl(sensor);

    constexpr double duration = 10.0;  // seconds

    std::vector<std::string> settings = {
        std::string("Stationary: ") + (is_stationary ? "true" : "false"),
        std::string("Output Data Rate: ") + std::to_string(odr_to_hz(sensor.odr)) + " Hz",
        std::string("Range: ") + std::to_string(range_to_g(sensor.g_range)) + " g"
    };

    auto samples  = read_continuous(sensor, duration);
    auto filename = write_to_csv(samples, settings);

    print_sample_preview(samples, 10);

    std::string sep(57, '=');
    std::cout << sep << '\n'
              << "Measurement complete!\n"
              << "Data saved to: " << filename << '\n'
              << sep << "\n\n";

    terminate(sensor);
}
